### ISSUE TITLE
BUG-000134502 ArcGIS World Geocoding Service does not provide proper suggestions in Crowdsource Reporter apps

### ISSUE THREAD

**Issue by aparveen** [Esri/crowdsource-reporter#497]
- Search addresses with ArcGIS World Geocoding Service in Configurable Apps Basic Viewer template, Crowdsource Reporter, Crowdsource Polling, and Web AppBuilder
- Only in the Crowdsource Reporter apps, the ArcGIS World Geocoding Service does not provide proper suggestions even the search is set to be within the map.
- Crowdsource Reporter apps: https://ess.maps.arcgis.com/home/item.html?id=3844ae1bd0a541ef867768f445cfd6d0

**Comment by aparveen** [comment]
**Reporter**
![CrowdsourceReporter_suggestions](https://user-images.githubusercontent.com/10148401/108744083-5e6a5f00-74ee-11eb-8bf9-c810fd451617.JPG)
**Basic Viewer**
![BasicViewer_Suggestions](https://user-images.githubusercontent.com/10148401/108744215-878aef80-74ee-11eb-94c6-b26aa299e415.jpg)

**Comment by ashishchoure** [comment]
@aparveen 

- We are using the custom search widget in the crowd source reporter application.
- This is the reason the search results are different in the Reporter app and in the Basic Viewer app.
- We think that changing the custom select tool will have a high impact and high LOE and it may impact the following components : 
    - Map Search
    - Geo Form
    - Layer Search (We use the custom search to directly select the feature from the map and it loads in the feature details)
    - UI of Search (We show results of all the configured geocoders which can be expanded and collapsed)
    - Accessibility
- In addition to the above points, we will have to consider different workflows which may get affected by this change.

Please let us know your thoughts.

**Comment by aparveen** [comment]
@ashishchoure please proceed with this change, could you please make this one separate delivery from the other fixes, so if there is any issue with the code we can pull it back?

**Comment by ashishchoure** [comment]
@aparveen 

- We are looking at all the possible solutions to make sure we have the minimum impact from this change.
- We will update you once we have a feasible approach.

**Comment by ashishchoure** [comment]
We have analyzed the issue and here are the two approaches: 
### 1. Replace the custom Search with ArcGIS JS API Search widget
   - **Pros :**
      - The result will exactly match with the applications which are using the search widget (Result may vary based on the locator settings)
   - **Cons :**
      - The API search widget will not show suggestions for the locator which do not have suggest capability
      - The UI will not match with the reporter application
         - Need to override the classes to match the UI with the custom search
         - Need to add different classes while showing the locator on the main screen and on the geo form screen as the UI is littlebit different.
      - Custom locator shows the addresses for all the configured locators in a section, and the API locator widget shows the addresses one after the other
      - This approach will have a very high impact and high LOE as we need to completely replace the custome search widget.


### 2. Add suggestion capabilities inside the custom Search widget
   - **Pros :**
      - No UI changes in the application
      - The result will match with the API’s locator widget. (Result may vary based on the different parameters.)
      - The impact and LOE will be less as compared to Approach 1
      - Backward Compatibility: 
         - We will show the addresses for geocoders that do not support suggest capability, as this was designed previously.
         - So, in the case of multiple geocoders, those who support suggestion capability will show the suggestions and others without suggest capability will directly show the addresses in the list
         
   - **Cons :**
      - As this a custom search widget this will not be in-line with the API search widget enhancements. 

**Additonal Notes :** 
- We are planning to honor all the search parameters added in the configuration which includes :
   - Maximum number of suggestions
   - Search with-in map extent
   - Enable suggestions : If the locator has a suggest capability and enable suggestions are turned on then only we will show the suggestions, otherwise we will show the addresses (This will handle the backward compatibility)



As per our discussion on yesterday's call, we decided to go ahead with the Apporach 2. This will have a minium impact on the application.

Please let us know your thoughts.

**Comment by ashishchoure** [comment]
### Dev Checklist

Items ​ | Is required​ | Status | Comments​
-- | -- | -- | --
Required NLS change?​ | No |   | ​
Required Backward compatibility​? | No |   |  
Required 508​? | No |   |  
Need Sanitize?​ | No

### Impact Analysis Report 
Impacted Areas | Comments
-- | --
Have Configuration changes?​ | No
Have Runtime changes?​ | Yes
Impact on RTL | No
Geocoding Service With Suggest Capability | <li>All the geocoding services with suggest capability should show suggestions in the collapsable list </li><li> The configurable parameter such as name, maximum suggestions, search within map extent, enable suggestions should be honored while showing suggestions </li><li>Correct address should be shown on the map once a user clicks on any of the suggestion</li>
Geocoding Service Without  Suggest Capability | <li>Geocoder without suggest capability should continue to work the same ways as per the current implementation</li>
Layer Search | <li>If any of the layers is search enabled then the layer search should continue to work the same ways as per the current implementation and the feature should be shown in the list</li>

**Comment by aparveen** [comment]
@ashishchoure in Android phone and table, the suggestion is not working as expected. If I start to type, suggestions do not show up right away, after finish typing, if I delete a few alphabets, then it starts to show up. Could you please take a look?

**Comment by ashishchoure** [comment]
@aparveen 

- This is an issue with Andriod devices. However, this is an existing issue in the application and we have not added any code which introduces this behavior.
- The keypress events do not work as expected in Andriod devices. We found a lot of content that describes a similar issue and [here ](https://stackoverflow.com/questions/22473950/keypress-event-not-firing-in-android-mobile) is one of them.
- We tried several approaches, but at this point in time, we don't see a possible solution to this behavior.
